opensearch_py_ml/semantic search/train.py

```diff
@@ -0,0 +1,362 @@
+
+            out_q = model.tokenize(batch_q)
+            for key in out_q.keys():
+                out_q[key] = out_q[key].to(device)
+            Y = model(out_q)["sentence_embedding"]
+
+            out = model.tokenize(batch)
+            for key in out.keys():
+                out[key] = out[key].to(device)
+            X = model(out)["sentence_embedding"]
+
+            XY = torch.exp(torch.matmul(X, Y.T) / (X.shape[1]) ** 0.5)
+            num = torch.diagonal(XY)
+
+            den0 = torch.sum(XY, dim=0)
+            den1 = torch.sum(XY, dim=1)
+
+            train_loss = -torch.sum(torch.log(num / den0)) - torch.sum(
+                torch.log(num / den1)
+            )
+            loss.append(train_loss.item())
+            train_loss.backward()
+            optimizer.step()
+            scheduler.step()
+            optimizer.zero_grad()
+
+            if not j % 500 and j != 0:  # MS: change 500
+                plt.plot(loss[::100])
+                plt.show()
+
+    # saving the pytorch model and the tokenizers.json file is saving at this step
+    model.save("trained_pytorch_model/")
+    cpu_model = model.to(device)
+    print(f" Total training time: {time.time() - init_time}")
+
+    out_q = model.tokenize(batch_q)
+    for key in out_q.keys():
+        out_q[key] = out_q[key].to(device)
+    traced_cpu = torch.jit.trace(
+        cpu_model,
+        ({"input_ids": out_q["input_ids"], "attention_mask": out_q["attention_mask"]}),
+        strict=False,
+    )
+
+    print("Preparing model to save")
+    torch.jit.save(traced_cpu, output_model_path)
+
+    print("Model saved to path: " + output_model_path)
+    return model
+
+
+def zip_model(model_path: str = None, zip_file_name: str = None) -> None:
+    """
+    Description:
+    zip the model file and its tokenizer.json file to prepare to upload to the Open Search cluster
+
+    Parameters:
+    model_path: str
+        Optional, path to find the model file, if None, default as trained_model.pt file in current path
+    zip_file_name: str =None
+        Optional, file name for zip file. if None, default as zip_model.zip
+
+    Return:
+        None
+    """
+
+    if model_path is None:
+        model_path = os.path.join(os.getcwd(), "trained_model.pt")
+
+    if zip_file_name is None:
+        zip_file_name = "zip_model.zip"
+
+    # Create a ZipFile Object
+    with ZipFile("zip_model.zip", "w") as zipObj:
+        zipObj.write(model_path)
+        zipObj.write("trained_pytorch_model/tokenizer.json")
+    print("zip file is saved to" + os.getcwd() + "/" + zip_file_name)
+
+
+def qryrem(x):
+    #         for removing the "QRY:" token
+    y = x.passages.split(" QRY:")[0].split("<|startoftext|>")
+    return y[1]
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Please enter argument to start semantic search training."
+    )
+    parser.add_argument("--read_path", type=str, default=None)
```

Comment:
I'm not sure, how are we expecting customers to execute this function? from the command terminal?

Reply:
yes it can be executive in command line  like this way python 3 train.py --read_path = '/home/file.zip'

Reply:
so if customer installs this package like other packages `panda`, `numpy`, do you expect customers to run a file like this way? Try to think like this is a python library customer will install and then import any function just like what we did for model upload:

```
from opensearch_py_ml.ml_commons_integration import MLCommonClient

ml_client = MLCommonClient(client)
ml_client.put_model("/workplace/dhrubo/upload_content/all-MiniLM-L6-v2.zip", "/workplace/dhrubo/upload_content/model_config.json", verbose=True)
```


I would suggest you to write the code this way.

```

Reply:
yes both ways are supported

Reply:
we can do 

`from opensearch_py_ml.semantic_search import train `
`training = train.train_semantic_search_model(read_path = '/home/ec2-user/opensearch-py-ml-dev/synthethic_queries.zip')`

Reply:
> yes it can be executive in command line like this way python 3 train.py --read_path = '/home/file.zip'

Not python exports, but looks like this is development/test way for developer, not a production way for the end user. In my mind, if we say we are building command line tool, it will be some shell command like [`find`](https://en.wikipedia.org/wiki/Find_(Unix)) command. 

For example, we build `opensearch-py-ml-cli` tool and user can install this command with `pip3 install opensearch-py-ml-cli`, then user can train model with these shell command `os-py-ml-train <huggingface_model_id> <sythentic_query_folder> <other params like num_epochs >`.

Reply:
Aligned to remove the main function to get parameters from command line, Considering for future development for commands